File: ch1/str_comp.cc
// str_comp takes a string and compresses the string using counts of repeated
//     characters. That is, aabcccccaa would be a2b1c5a3
//     If the compressed string is not smaller than the original string, then
//     return the original string

#include <iostream>
#include <string>

std::string str_comp(std::string &str) {
  // Getting length
  int orig = str.length();
  if (orig <= 1) return str;

  // Setting up iteration
  std::string result = "";
  char last = str[0];
  int count = 1;

  for (int i = 1; i < orig; i++) {
    if (str[i] == last) {
      count++;
    }

    else {
      // Convert character to string and add to result
      std::string charstr(1, last);
      result += charstr;

      // Convert the count to string and add to result
      std::string numstr = std::to_string(count);
      result += numstr;

      // Set new char and count for iteration
      last = str[i];
      count = 1;
    }
  }

  // We have to add the last character(s)
  std::string charstr(1, last);
  result += charstr;

  std::string numstr = std::to_string(count);
  result += numstr;

  // Check to ensure compression actually compressed it
  int new_len = result.length();

  if (new_len < orig) return result;
  else return str;
}

int main() {
  std::string test = "aabcccccaaa";
  std::string res = str_comp(test);

  std::cout << res << std::endl;
}

File: ch1/str_comp.py
def str_comp(str1):
    len1 = len(str1)
    if (not len1):
        return str1

    last_char = str1[0]
    count = 1
    result = ''

    for i in range(1, len1):
        if str1[i] != last_char:
            result += last_char + str(count)
            last_char = str1[i]
            count = 1

        else:
            count += 1

    result += last_char + str(count)

    if (len(result) >= len1):
        return str1
    else: 
        return result



def main():
    str1 = "aaa"
    str2 = "ABCCBA"
    str3 = "aabcccccaaa"

    print(str_comp(str1))
    print(str_comp(str2))
    print(str_comp(str3))

if __name__ == "__main__":
    main()